File: C/Ordenaciones/MergeSort.c
#include <stdio.h>
#include <stdlib.h>

// Funcion que ordena 2 subarrays
// First subarray is arr[l..m]
// Second subarray is arr[m+1..r]
// Complejidad temporal: O(n log n)
// Complejidad espacial: O(n)
void merge(int a[], int izq, int m, int der, int n) {       
    int i=izq, j=m+1, k=izq;    
    int aux[n];
    for (i=izq; i<=der; i++) aux[i]=a[i];
    
    i=izq;     
    while (i<=m && j<=der) {                                     
        if (aux[i]<=aux[j]) a[k++]=aux[i++];
        else a[k++]=aux[j++];
    }

    // copia los elementos que quedan de la primera mitad (si los hay)
    while (i<=m) { a[k++]=aux[i++]; } 
    while (j<=der) { a[k++]=aux[j++]; } 
}
void mergeSort(int a[], int izq, int der, int n) {
    if (izq<der) {        
        int m=(izq+der)/2; // Mitad del array actual        
        mergeSort(a, izq, m, n); // Ordena la mitad 1
        mergeSort(a, m+1, der, n); // Ordena la mitad 2      
        merge(a, izq, m, der, n);  // Ordena las mitades ordenadas
    }
}

File: PYTHON/Ordenaciones/MergeSort.py
class MergeSort:
    def merge(self, a, izq, m, der):
        i = izq
        j = m + 1
        k = izq
        aux = [0] * len(a)
        for i in range(izq, der+1):
            aux[i] = a[i]

        i = izq

        while i <= m and j <= der:
            if aux[i] <= aux[j]:
                a[k] = aux[i]
                i += 1
            else:
                a[k] = aux[j]
                j += 1
            k += 1

        while i <= m:
            a[k] = aux[i]
            k += 1
            i += 1

        while j <= der:
            a[k] = aux[j]
            k += 1
            j += 1

    def mergesort(self, a, izq, der):
        if izq < der:
            m = (izq + der) // 2
            self.mergesort(a, izq, m)
            self.mergesort(a, m+1, der)
            self.merge(a, izq, m, der)
